### PR TITLE
HIVE-26438: Remove unnecessary optimization in canHandleQbForCbo()

### PR DESCRIPTION
**Pull request by achennagiri**
### What changes were proposed in this pull request?
This ticket is an improvement on https://issues.apache.org/jira/browse/HIVE-26426.  The canHandleQbForCbo() checks whether Calcite can handle the query or not and it returns null if the query can be handled; non-null reason string if it cannot be. 

But currently, it returns an empty string if INFO Log is not enabled which should not be the case.
Also, this is probably a performance optimization that is not needed and can be simplified. 

### Why are the changes needed?
Simplify the code and improve the correctness.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 1]
```diff
@@ -987,30 +985,29 @@ private static String canHandleQbForCbo(QueryProperties queryProperties, HiveCon
...
-        msg += "uses scripts; ";
-      }
-      if (queryProperties.hasLateralViews()) {
-        msg += "has lateral views; ";
-      }
-      if (msg.isEmpty()) {
-        msg += "has some unspecified limitations; ";
-      }
-      msg = msg.substring(0, msg.length() - 2);
+    if (queryProperties.hasClusterBy()) {
```
Comment (jfsii) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:988]: I would maybe change this to something like (but this is more of an opinion):
List<String> reasons = new ArrayList();
and in each if () { reason.add(errorMsg); }
and at the end do:
return Strings.join("; ", reasons);

You could technically combine the top if check for validity with the error message processing to force each case to always have a valid error message by deleting the top if (conditions) { return null },
add the missing condition to the error messages (queryProperties.isCBOSupportedLateralViews())) and change the last return to be something like:
return reasons.isEmpty() ? null : Strings.join("; ",  reasons);

Hopefully that makes sense? It is also a matter of style - feel free to ignore if you do not think the above suggestions make the code more readable or less fragile. (I'm not 100% sure they make it better - but in my head I think they do).

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 2]
```diff
@@ -987,30 +985,29 @@ private static String canHandleQbForCbo(QueryProperties queryProperties, HiveCon
...
+      msg += "has PTF; ";
     }
+    if (queryProperties.usesScript()) {
+      msg += "uses scripts; ";
+    }
+    if (queryProperties.hasLateralViews()) {
+      msg += "has lateral views; ";
+    }
+    if (msg.isEmpty()) {
+    msg += "has some unspecified limitations; ";
```
Comment (jfsii) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:1007]: indenting is incorrect  -
one positive change from my suggestions above is that it enforces anyone who modifies this section of code with new conditions - would be forced to also provide a valid error message (so the only time msgs/reason would/should be empty is if it is a valid QB for  CBO).

**Review by jfsii — COMMENTED** [review]
Thanks for the follow up work. I left a couple of notes .

**Review by zabetak — COMMENTED** [review]
I agree to everything @jfsii said; let's incorporate their suggestions and I will merge this when tests come out green.

nit: I would prefer to use `String.join` instead of `Strings.join` to reduce dependencies to other libraries.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java [hunk 3]
```diff
@@ -974,44 +972,40 @@ Pair<Boolean, String> canCBOHandleAst(ASTNode ast, QB qb, PreCboCtx cboCtx) {
...
+    }
+    if (queryProperties.usesScript()) {
+      errorMsg = "uses scripts";
+      reasons.add(errorMsg);
+    }
+    if (queryProperties.hasLateralViews()) {
+      errorMsg = "has lateral views";
+      reasons.add(errorMsg);
+    }
+    if (!queryProperties.isCBOSupportedLateralViews()) {
```
Comment (achennagiri) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:1004]: Trying to keep the error messaging the same. If isCBOSupportedLateralViews() returns False, only then the error message will be "has some unspecified limitations".

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 1 above)
Comment (achennagiri) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:980]: @jfsii I have made the changes accordingly. Let me know if there are any issues

File: ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (achennagiri) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java:1004]: I see! Thanks a lot for this review. 
I agree `"has some unspecified limitations"` is not reachable.  I will omit it.

**Review by zabetak — APPROVED** [review]
LGTM!

**Pull request closed without merging**